File: src/data/courses/dsa/two-pointers-sliding-window/two-pointers-same-direction/long-pressed-name/code/python/long_pressed_name.py
# ============================================================
# Long Pressed Name
# ============================================================

# ============================================================
# Solution: Two Pointers
# Time: O(N+M) | Space: O(1)
# ============================================================
class Solution:
    def isLongPressedName(self, name: str, typed: str) -> bool:
        i = j = 0

        while j < len(typed):
            if i < len(name) and name[i] == typed[j]:
                i += 1  # Normal match
                j += 1
            elif j > 0 and typed[j] == typed[j - 1]:
                j += 1  # Long press — skip
            else:
                return False  # Mismatch

        return i == len(name)  # All name characters consumed?

File: src/data/courses/dsa/two-pointers-sliding-window/two-pointers-same-direction/long-pressed-name/code/java/LongPressedName.java
// ============================================================
// Long Pressed Name
// ============================================================

// ============================================================
// Solution: Two Pointers
// Time: O(N+M) | Space: O(1)
// ============================================================
class Solution {
    public boolean isLongPressedName(String name, String typed) {
        int i = 0, j = 0;

        while (j < typed.length()) {
            if (i < name.length() && name.charAt(i) == typed.charAt(j)) {
                i++; j++;
            } else if (j > 0 && typed.charAt(j) == typed.charAt(j - 1)) {
                j++;  // Long press
            } else {
                return false;
            }
        }

        return i == name.length();
    }
}
